File: src/schemas/pattern/pattern.ts
import { field, fixedArray, option, vec } from "@dao-xyz/borsh";

export class PatternProperties {
  @field({ type: "u16" })
  width: number;

  @field({ type: "u16" })
  height: number;

  constructor(data: PatternProperties) {
    this.width = data.width;
    this.height = data.height;
  }
}

export class PatternInfo {
  @field({ type: "string" })
  title: string;

  @field({ type: "string" })
  author: string;

  @field({ type: "string" })
  company: string;

  @field({ type: "string" })
  copyright: string;

  @field({ type: "string" })
  description: string;

  constructor(data: PatternInfo) {
    this.title = data.title;
    this.author = data.author;
    this.company = data.company;
    this.copyright = data.copyright;
    this.description = data.description;
  }
}

export class PaletteItemStitchStrands {
  @field({ type: option("u8") })
  full?: number;

  @field({ type: option("u8") })
  petite?: number;

  @field({ type: option("u8") })
  half?: number;

  @field({ type: option("u8") })
  quarter?: number;

  @field({ type: option("u8") })
  back?: number;

  @field({ type: option("u8") })
  straight?: number;

  @field({ type: option("u8") })
  frenchKnot?: number;

  @field({ type: option("u8") })
  special?: number;

  constructor(data: PaletteItemStitchStrands) {
    this.full = data.full;
    this.petite = data.petite;
    this.half = data.half;
    this.quarter = data.quarter;
    this.back = data.back;
    this.straight = data.straight;
    this.frenchKnot = data.frenchKnot;
    this.special = data.special;
  }
}

export class Blend {
  @field({ type: "string" })
  brand: string;

  @field({ type: "string" })
  number: string;

  @field({ type: "u8" })
  strands: number;

  constructor(data: Blend) {
    this.brand = data.brand;
    this.number = data.number;
    this.strands = data.strands;
  }
}

export class Bead {
  @field({ type: "f32" })
  length: number;

  @field({ type: "f32" })
  diameter: number;

  constructor(data: Bead) {
    this.length = data.length;
    this.diameter = data.diameter;
  }
}

export class PaletteItem {
  @field({ type: "string" })
  brand: string;

  @field({ type: "string" })
  number: string;

  @field({ type: "string" })
  name: string;

  @field({ type: "string" })
  color: string;

  @field({ type: option(vec(Blend)) })
  blends?: Blend[];

  @field({ type: option(Bead) })
  bead?: Bead;

  @field({ type: option(PaletteItemStitchStrands) })
  strands?: PaletteItemStitchStrands;

  constructor(data: PaletteItem) {
    this.brand = data.brand;
    this.number = data.number;
    this.name = data.name;
    this.color = data.color;
    this.blends = data.blends;
    this.bead = data.bead;
    this.strands = data.strands;
  }
}

export type StitchesPerInch = [number, number];
export class Fabric {
  @field({ type: fixedArray("u16", 2) })
  spi: StitchesPerInch;

  @field({ type: "string" })
  kind: string;

  @field({ type: "string" })
  name: string;

  @field({ type: "string" })
  color: string;

  constructor(data: Fabric) {
    this.spi = data.spi;
    this.kind = data.kind;
    this.name = data.name;
    this.color = data.color;
  }
}

export class FullStitch {
  @field({ type: "f32" })
  x: number;

  @field({ type: "f32" })
  y: number;

  @field({ type: "u8" })
  palindex: number;

  @field({
    serialize: (kind, writer) => (kind === FullStitchKind.Full ? writer.u8(0) : writer.u8(1)),
    deserialize: (reader) => (reader.u8() === 0 ? FullStitchKind.Full : FullStitchKind.Petite),
  })
  kind: FullStitchKind;

  constructor(data: FullStitch) {
    this.x = data.x;
    this.y = data.y;
    this.palindex = data.palindex;
    this.kind = data.kind;
  }
}

export const enum FullStitchKind {
  Full = "Full",
  Petite = "Petite",
}

export class PartStitch {
  @field({ type: "f32" })
  x: number;

  @field({ type: "f32" })
  y: number;

  @field({ type: "u8" })
  palindex: number;

  @field({
    serialize: (direction, writer) => (direction === PartStitchDirection.Forward ? writer.u8(1) : writer.u8(2)),
    deserialize: (reader) => (reader.u8() === 1 ? PartStitchDirection.Forward : PartStitchDirection.Backward),
  })
  direction: PartStitchDirection;

  @field({
    serialize: (kind, writer) => (kind === PartStitchKind.Half ? writer.u8(0) : writer.u8(1)),
    deserialize: (reader) => (reader.u8() === 0 ? PartStitchKind.Half : PartStitchKind.Quarter),
  })
  kind: PartStitchKind;

  constructor(data: PartStitch) {
    this.x = data.x;
    this.y = data.y;
    this.palindex = data.palindex;
    this.direction = data.direction;
    this.kind = data.kind;
  }
}

export const enum PartStitchDirection {
  Forward = "Forward",
  Backward = "Backward",
}

export const enum PartStitchKind {
  Half = "Half",
  Quarter = "Quarter",
}

export class LineStitch {
  @field({ type: fixedArray("f32", 2) })
  x: [number, number];

  @field({ type: fixedArray("f32", 2) })
  y: [number, number];

  @field({ type: "u8" })
  palindex: number;

  @field({
    serialize: (kind, writer) => (kind === LineStitchKind.Back ? writer.u8(0) : writer.u8(1)),
    deserialize: (reader) => (reader.u8() === 0 ? LineStitchKind.Straight : PartStitchKind.Quarter),
  })
  kind: LineStitchKind;

  constructor(data: LineStitch) {
    this.x = data.x;
    this.y = data.y;
    this.palindex = data.palindex;
    this.kind = data.kind;
  }
}

export const enum LineStitchKind {
  Back = "Back",
  Straight = "Straight",
}

export class NodeStitch {
  @field({ type: "f32" })
  x: number;

  @field({ type: "f32" })
  y: number;

  @field({ type: "bool" })
  rotated: boolean;

  @field({ type: "u8" })
  palindex: number;

  @field({
    serialize: (kind, writer) => (kind === NodeStitchKind.FrenchKnot ? writer.u8(0) : writer.u8(1)),
    deserialize: (reader) => (reader.u8() === 0 ? NodeStitchKind.FrenchKnot : NodeStitchKind.Bead),
  })
  kind: NodeStitchKind;

  constructor(data: NodeStitch) {
    this.x = data.x;
    this.y = data.y;
    this.palindex = data.palindex;
    this.rotated = data.rotated;
    this.kind = data.kind;
  }
}

export const enum NodeStitchKind {
  FrenchKnot = "FrenchKnot",
  Bead = "Bead",
}

export class CurvedStitch {
  @field({ type: vec(fixedArray("f32", 2)) })
  points: [number, number][];

  constructor(data: CurvedStitch) {
    this.points = data.points;
  }
}

export class SpecialStitch {
  @field({ type: "f32" })
  x: number;

  @field({ type: "f32" })
  y: number;

  @field({ type: "u16" })
  rotation: number;

  @field({ type: fixedArray("bool", 2) })
  flip: [boolean, boolean];

  @field({ type: "u8" })
  palindex: number;

  @field({ type: "u8" })
  modindex: number;

  constructor(data: SpecialStitch) {
    this.x = data.x;
    this.y = data.y;
    this.palindex = data.palindex;
    this.modindex = data.modindex;
    this.rotation = data.rotation;
    this.flip = data.flip;
  }
}

export class SpecialStitchModel {
  @field({ type: "string" })
  uniqueName: string;

  @field({ type: "string" })
  name: string;

  @field({ type: vec(NodeStitch) })
  nodes: NodeStitch[];

  @field({ type: vec(LineStitch) })
  lines: LineStitch[];

  @field({ type: vec(CurvedStitch) })
  curves: CurvedStitch[];

  constructor(data: SpecialStitchModel) {
    this.uniqueName = data.uniqueName;
    this.name = data.name;
    this.nodes = data.nodes;
    this.lines = data.lines;
    this.curves = data.curves;
  }
}

export class Pattern {
  @field({ type: PatternProperties })
  properties: PatternProperties;

  @field({ type: PatternInfo })
  info: PatternInfo;

  @field({ type: vec(PaletteItem) })
  palette: PaletteItem[];

  @field({ type: Fabric })
  fabric: Fabric;

  @field({ type: vec(FullStitch) })
  fullstitches: FullStitch[];

  @field({ type: vec(PartStitch) })
  partstitches: PartStitch[];

  @field({ type: vec(NodeStitch) })
  nodes: NodeStitch[];

  @field({ type: vec(LineStitch) })
  lines: LineStitch[];

  @field({ type: vec(SpecialStitch) })
  specialstitches: SpecialStitch[];

  @field({ type: vec(SpecialStitchModel) })
  specialStitchModels: SpecialStitchModel[];

  constructor(data: Pattern) {
    this.properties = data.properties;
    this.info = data.info;
    this.palette = data.palette;
    this.fabric = data.fabric;
    this.fullstitches = data.fullstitches;
    this.partstitches = data.partstitches;
    this.nodes = data.nodes;
    this.lines = data.lines;
    this.specialstitches = data.specialstitches;
    this.specialStitchModels = data.specialStitchModels;
  }
}

export type Stitch = { full: FullStitch } | { part: PartStitch } | { node: NodeStitch } | { line: LineStitch };
export type StitchKind = FullStitchKind | PartStitchKind | NodeStitchKind | LineStitchKind;

File: src-tauri/src/core/pattern/pattern.rs
use borsh::{BorshDeserialize, BorshSerialize};
use serde::{Deserialize, Serialize};

use super::stitches::*;

#[derive(Debug, Default, Clone, BorshSerialize, BorshDeserialize)]
pub struct Pattern {
  pub properties: PatternProperties,
  pub info: PatternInfo,
  pub palette: Vec<PaletteItem>,
  pub fabric: Fabric,
  pub fullstitches: Stitches<FullStitch>,
  pub partstitches: Stitches<PartStitch>,
  pub nodes: Stitches<Node>,
  pub lines: Stitches<Line>,
  pub specialstitches: Stitches<SpecialStitch>,
  pub special_stitch_models: Vec<SpecialStitchModel>,
}

impl Pattern {
  /// Check if the pattern contains a stitch.
  pub fn contains_stitch(&self, stitch: &Stitch) -> bool {
    match stitch {
      Stitch::Full(fullstitch) => self.fullstitches.contains(fullstitch),
      Stitch::Part(partstitch) => self.partstitches.contains(partstitch),
      Stitch::Node(node) => self.nodes.contains(node),
      Stitch::Line(line) => self.lines.contains(line),
    }
  }

  /// Adds many stitches to the pattern.
  pub fn add_stitches(&mut self, stitches: Vec<Stitch>) {
    for stitch in stitches {
      self.add_stitch(stitch);
    }
  }

  /// Adds a stitch to the pattern and returns any conflicts that may have arisen.
  pub fn add_stitch(&mut self, stitch: Stitch) -> Vec<Stitch> {
    log::trace!("Adding stitch");
    let mut conflicts = Vec::new();
    match stitch {
      Stitch::Full(fullstitch) => {
        match fullstitch.kind {
          FullStitchKind::Full => {
            conflicts.extend(
              self
                .fullstitches
                .remove_conflicts_with_full_stitch(&fullstitch)
                .into_iter()
                .map(Stitch::Full),
            );
            conflicts.extend(
              self
                .partstitches
                .remove_conflicts_with_full_stitch(&fullstitch)
                .into_iter()
                .map(Stitch::Part),
            );
          }
          FullStitchKind::Petite => {
            conflicts.extend(
              self
                .fullstitches
                .remove_conflicts_with_petite_stitch(&fullstitch)
                .into_iter()
                .map(Stitch::Full),
            );
            conflicts.extend(
              self
                .partstitches
                .remove_conflicts_with_petite_stitch(&fullstitch)
                .into_iter()
                .map(Stitch::Part),
            );
          }
        };
        if let Some(fullstitch) = self.fullstitches.insert(fullstitch) {
          conflicts.push(Stitch::Full(fullstitch));
        }
      }
      Stitch::Part(partstitch) => {
        match partstitch.kind {
          PartStitchKind::Half => {
            conflicts.extend(
              self
                .fullstitches
                .remove_conflicts_with_half_stitch(&partstitch)
                .into_iter()
                .map(Stitch::Full),
            );
            conflicts.extend(
              self
                .partstitches
                .remove_conflicts_with_half_stitch(&partstitch)
                .into_iter()
                .map(Stitch::Part),
            );
          }
          PartStitchKind::Quarter => {
            conflicts.extend(
              self
                .fullstitches
                .remove_conflicts_with_quarter_stitch(&partstitch)
                .into_iter()
                .map(Stitch::Full),
            );
            conflicts.extend(
              self
                .partstitches
                .remove_conflicts_with_quarter_stitch(&partstitch)
                .into_iter()
                .map(Stitch::Part),
            );
          }
        };
        if let Some(partstitch) = self.partstitches.insert(partstitch) {
          conflicts.push(Stitch::Part(partstitch));
        }
      }
      Stitch::Node(node) => {
        if let Some(node) = self.nodes.insert(node) {
          conflicts.push(Stitch::Node(node));
        }
      }
      Stitch::Line(line) => {
        if let Some(line) = self.lines.insert(line) {
          conflicts.push(Stitch::Line(line));
        }
      }
    };
    conflicts
  }

  /// Removes many stitches from the pattern.
  pub fn remove_stitches(&mut self, stitches: Vec<Stitch>) {
    for stitch in stitches {
      self.remove_stitch(stitch);
    }
  }

  /// Removes and returns a stitch from the pattern.
  pub fn remove_stitch(&mut self, stitch: Stitch) -> Option<Stitch> {
    log::trace!("Removing stitch");
    match stitch {
      Stitch::Full(fullstitch) => self.fullstitches.remove(&fullstitch).map(|fs| fs.into()),
      Stitch::Part(partstitch) => self.partstitches.remove(&partstitch).map(|ps| ps.into()),
      Stitch::Node(node) => self.nodes.remove(&node).map(|node| node.into()),
      Stitch::Line(line) => self.lines.remove(&line).map(|line| line.into()),
    }
  }

  /// Removes and returns all stitches with a given palette index from the pattern.
  pub fn remove_stitches_by_palindex(&mut self, palindex: u8) -> Vec<Stitch> {
    log::trace!("Removing stitches by palette index");
    let mut conflicts = Vec::new();
    conflicts.extend(
      self
        .fullstitches
        .remove_stitches_by_palindex(palindex)
        .into_iter()
        .map(Stitch::Full),
    );
    conflicts.extend(
      self
        .partstitches
        .remove_stitches_by_palindex(palindex)
        .into_iter()
        .map(Stitch::Part),
    );
    conflicts.extend(
      self
        .lines
        .remove_stitches_by_palindex(palindex)
        .into_iter()
        .map(Stitch::Line),
    );
    conflicts.extend(
      self
        .nodes
        .remove_stitches_by_palindex(palindex)
        .into_iter()
        .map(Stitch::Node),
    );
    conflicts
  }

  pub fn restore_stitches(&mut self, stitches: Vec<Stitch>, palindex: u8) {
    let mut fullstitches = Vec::new();
    let mut partstitches = Vec::new();
    let mut lines = Vec::new();
    let mut nodes = Vec::new();
    for stitch in stitches.into_iter() {
      match stitch {
        Stitch::Full(fullstitch) => fullstitches.push(fullstitch),
        Stitch::Part(partstitch) => partstitches.push(partstitch),
        Stitch::Line(line) => lines.push(line),
        Stitch::Node(node) => nodes.push(node),
      }
    }

    self.fullstitches.restore_stitches(fullstitches, palindex);
    self.partstitches.restore_stitches(partstitches, palindex);
    self.lines.restore_stitches(lines, palindex);
    self.nodes.restore_stitches(nodes, palindex);
  }
}

#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize)]
pub struct PatternProperties {
  pub width: u16,
  pub height: u16,
}

impl Default for PatternProperties {
  fn default() -> Self {
    Self { width: 100, height: 100 }
  }
}

#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize)]
pub struct PatternInfo {
  pub title: String,
  pub author: String,
  pub company: String,
  pub copyright: String,
  pub description: String,
}

impl Default for PatternInfo {
  fn default() -> Self {
    Self {
      title: String::from("Untitled"),
      author: String::new(),
      company: String::new(),
      copyright: String::new(),
      description: String::new(),
    }
  }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, BorshSerialize, BorshDeserialize)]
pub struct PaletteItem {
  pub brand: String,
  pub number: String,
  pub name: String,
  pub color: String,
  pub blends: Option<Vec<Blend>>,
  pub bead: Option<Bead>,
  pub strands: Option<PaletteItemStitchStrands>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, BorshSerialize, BorshDeserialize)]
pub struct Blend {
  pub brand: String,
  pub number: String,
  pub strands: BlendStrands,
}

#[nutype::nutype(
  sanitize(with = |raw| raw.clamp(1, 6)),
  derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize, BorshSerialize, BorshDeserialize)
)]
pub struct BlendStrands(u8);

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, BorshSerialize, BorshDeserialize)]
pub struct Bead {
  pub length: f32,
  pub diameter: f32,
}

#[derive(Debug, Default, Clone, PartialEq, Serialize, Deserialize, BorshSerialize, BorshDeserialize)]
pub struct StitchStrandsStruct<T> {
  pub full: T,
  pub petite: T,
  pub half: T,
  pub quarter: T,
  pub back: T,
  pub straight: T,
  pub french_knot: T,
  pub special: T,
}

#[nutype::nutype(
  sanitize(with = |raw| raw.clamp(1, 12)),
  derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize, BorshSerialize, BorshDeserialize)
)]
pub struct StitchStrands(u8);

pub type PaletteItemStitchStrands = StitchStrandsStruct<Option<StitchStrands>>;
pub type DefaultStitchStrands = StitchStrandsStruct<StitchStrands>;

impl Default for DefaultStitchStrands {
  fn default() -> Self {
    Self {
      full: StitchStrands::new(2),
      petite: StitchStrands::new(2),
      half: StitchStrands::new(2),
      quarter: StitchStrands::new(2),
      back: StitchStrands::new(1),
      straight: StitchStrands::new(1),
      french_knot: StitchStrands::new(2),
      special: StitchStrands::new(2),
    }
  }
}

#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize)]
pub struct Fabric {
  pub spi: StitchesPerInch,
  pub kind: String,
  pub name: String,
  pub color: String,
}

impl Default for Fabric {
  fn default() -> Self {
    Self {
      spi: (14, 14),
      kind: String::from("Aida"),
      name: String::from("White"),
      color: String::from("FFFFFF"),
    }
  }
}

pub type StitchesPerInch = (u16, u16);
